File: src/main/resources/templates/javascript/solution/src/context.js
export default class Context {
    /** @type {?{ performSort: (input: Date[]) => void }} */
    #sortAlgorithm = null;

    /** @type {Date[]} */
    #dates = [];

    /**
     * Runs the configured sort algorithm.
     */
    sort() {
        this.#sortAlgorithm?.performSort(this.#dates);
    }

    get sortAlgorithm() {
        return this.#sortAlgorithm;
    }

    set sortAlgorithm(sortAlgorithm) {
        this.#sortAlgorithm = sortAlgorithm;
    }

    get dates() {
        return this.#dates;
    }

    set dates(dates) {
        this.#dates = dates;
    }
}

File: src/main/resources/templates/dart/solution/lib/context.dart
import 'sort_strategy.dart';

class Context {
  List<DateTime>? dates;
  SortStrategy? sortAlgorithm;

  /// Runs the configured sort algorithm.
  void sort() {
    final dates = this.dates;
    if (dates == null) {
      throw StateError('dates has not been set');
    }

    final sortAlgorithm = this.sortAlgorithm;
    if (sortAlgorithm == null) {
      throw StateError('sortStrategy has not been set');
    }

    sortAlgorithm.performSort(dates);
  }
}
